### PR TITLE
[5] Add checkout-code step to build

### PR DESCRIPTION
**Pull request by minzcmu**
- Add checkout-code step to build

Blocked by https://github.com/screwdriver-cd/screwdriver/pull/356
Related to https://github.com/screwdriver-cd/screwdriver/issues/312

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -283,6 +283,8 @@ factory.create(config).then(model => {
 | config.username | String | Yes | User who made the change to kick off the build |
 | config.container | String | No | Container for the build to run in |
 | config.sha | String | No | SHA used to kick off the build |
+| config.prRef | String | No | PR branch or reference; required for PR jobs |
+| config.eventId | String | No | Id of the event this build belongs to |
```
Comment (tkyi) [inline README.md:287]: 👍  nice!

**Comment by petey** [comment]
Why are we renaming sd-setup?

File: test/lib/buildFactory.test.js [hunk 2]
```diff
@@ -156,6 +161,7 @@ describe('Build Factory', () => {
         beforeEach(() => {
             scmMock.getCommitSha.resolves(sha);
             scmMock.decorateCommit.resolves(commit);
+            scmMock.getCheckoutCommand.resolves({ name: 'checkout-code', command: 'git clone' });
```
Comment (tkyi) [inline test/lib/buildFactory.test.js]: could you use `resolves(steps[0])`?

**Review by tkyi — APPROVED** [review]
Nice!

**Comment by minzcmu** [comment]
@petey  In the launcher, we are doing `checkout`, `set up workspace`, and `set up environments` in `sd-setup` step. We just move the `checkout` part out here, so we think `checkout-code` is more accurate. Can we still keep `sd-setup` and log something about  `set up workspace`, and `set up environments` in that step? Is it better? Then `sd-setup` will be the first step, and `checkout-code` will be the second one.

File: lib/buildFactory.js [hunk 3]
```diff
@@ -136,29 +137,48 @@ class BuildFactory extends BaseFactory {
...
+                    }
+
+                    modelConfig.container = permutation.image;
+                    modelConfig.environment = permutation.environment;
+                    modelConfig.steps = permutation.commands.map(command => ({
+                        name: command.name,
+                        command: command.command
+                    }));
+
+                    const scmParts = pipeline.scmUri.split(':');
```
Comment (d2lam) [inline lib/buildFactory.js]: You can do this
`const [host,,branch] = pipeline.scmUri.split(':');`

File: lib/buildFactory.js [hunk 4]
```diff
@@ -136,29 +137,48 @@ class BuildFactory extends BaseFactory {
...
+                    const orgRepo = pipeline.scmRepo.name.split('/');
```
Comment (d2lam) [inline lib/buildFactory.js]: Same thing here:
`const [org,repo] = pipeline.scmUri.split('/');`

So there aren't too many magic numbers below.

File: lib/buildFactory.js [hunk 5]
```diff
@@ -136,29 +137,48 @@ class BuildFactory extends BaseFactory {
...
+                        repo: orgRepo[1],
+                        sha: modelConfig.sha
+                    };
+
+                    if (modelConfig.prRef) {
+                        checkoutConfig.prRef = modelConfig.prRef;
+                    }
+
+                    return this.scm.getCheckoutCommand(checkoutConfig).then((command) => {
+                        modelConfig.steps.unshift(command);
```
Comment (d2lam) [inline lib/buildFactory.js:176]: Don't we still need the `sd-setup` for other things?

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "screwdriver-models",
-  "version": "17.1.0",
+  "version": "17.2.0",
```
Comment (d2lam) [inline package.json]: I feel like this should be a major bump.

**Review by d2lam — CHANGES_REQUESTED** [review]

File: package.json (hunk 6 above)
Comment (petey) [reply package.json]: I agree that this is a major version bump.

File: test/lib/buildFactory.test.js [hunk 7]
```diff
@@ -116,12 +117,16 @@ describe('Build Factory', () => {
...
+            name: 'screwdriver-cd/models'
+        };
+        const prRef = 'pull/3/merge';
         const username = 'i_made_the_request';
         const dateNow = Date.now();
         const isoTime = (new Date(dateNow)).toISOString();
         const container = 'node:4';
         const steps = [
-            { name: 'sd-setup' },
+            { command: 'git clone', name: 'checkout-code' },
```
Comment (petey) [inline test/lib/buildFactory.test.js]: If this step is named something other than `sd-setup`, then it needs to start with `sd-` to keep from stepping on user configs. 

If the launcher is still doing other special commands in `sd-setup` maybe those should also be moved out of launcher and inserted by the models. Maybe that should be done in another PR. We want to keep the `sd-` steps to a minimum.

File: lib/buildFactory.js (hunk 3 above)
Comment (petey) [reply lib/buildFactory.js]: I wasn't aware of this feature of destructuring. Thanks!

**Review by petey — CHANGES_REQUESTED** [review]

File: lib/buildFactory.js (hunk 5 above)
Comment (minzcmu) [reply lib/buildFactory.js:176]: Add it back now.

File: test/lib/buildFactory.test.js (hunk 7 above)
Comment (minzcmu) [reply test/lib/buildFactory.test.js]: Makes sense to add `sd-` as the prefix! 👍 Keep the `sd-setup` for now and will see if we can move those commands out in another PR. Rename this to  `sd-checkout-code`. @tkyi We need to change scm-base, scm-bitbucket and scm-github with the step name  `sd-checkout-code`.

Comment (minzcmu) [reply test/lib/buildFactory.test.js]: @tkyi Or can we just add a prefix `sd-` to command.name here?

Comment (d2lam) [reply test/lib/buildFactory.test.js]: We can set them to `sd-checkout-code` to be consistent across. So we can tell user that those steps are reserved. If we just add prefix, it's harder to document which steps are reserved and should not be inside the user config.

Comment (tkyi) [reply test/lib/buildFactory.test.js]: @minz1027 okay I'll start opening pull requests for those 👍

File: lib/buildFactory.js [hunk 8]
```diff
@@ -136,29 +137,49 @@ class BuildFactory extends BaseFactory {
...
-                        modelConfig.steps = permutation.commands.map(command => ({
-                            name: command.name,
-                            command: command.command
-                        }));
+                return Promise.all([
+                    job.pipeline,
+                    getCommitInfo({ job, scm: this.scm, modelConfig })
+                ]).then(([pipeline, data]) => {
+                    // TODO: support matrix jobs
+                    const index = number.toString().split('.')[1] || 0;
```
Comment (FenrirUnbound) [inline lib/buildFactory.js:145]: Does `Date.now()` return a number with a decimal?

Comment (tkyi) [reply lib/buildFactory.js:145]: No. We're not supporting matrix jobs yet, so it'll always use index 0 right now and only use the first permutation object.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,10 @@
 # CHANGELOG
 
+## 18.0.0
```
Comment (FenrirUnbound) [inline CHANGELOG.md:3]: It's a major version bump, so this would indicate that there is a backwards-incompatible change. Could we point out what's the breaking change in this release?

**Review by FenrirUnbound — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 9 above)
Comment (tkyi) [reply CHANGELOG.md:3]: Added a `breaking changes` section

**Review by FenrirUnbound — APPROVED** [review]
👍 

insane teeterboard skills - imgur

**Review by d2lam — APPROVED** [review]
Looks good 👍
